[RFC][mlir] Conditional support for fast-math attributes.

vzakhari opened the pull request:
This patch suggests changes for operations that support
arith::ArithFastMathInterface/LLVM::FastmathFlagsInterface.
Some of the operations may have fast-math flags not equal to `none`
only if they operate on floating point values.

This is inspired by https://llvm.org/docs/LangRef.html#fastmath-return-types
and my goal to add fast-math support for `arith.select` operation
that may produce results of any type.

The changes add new isArithFastMathApplicable/isFastmathApplicable
methods to the above interfaces that tell whether an operation
supporting the interface may have non-none fast-math flags.

LLVM dialect isFastmathApplicable implementation is based on https://github.com/llvm/llvm-project/blob/bac62ee5b473e70981a6bd9759ec316315fca07d/llvm/include/llvm/IR/Operator.h#L380
ARITH dialect isArithFastMathApplicable is more relaxed, because
it has to support custom MLIR types. This is the area where
improvements are needed (see TODO comments). I will appreciate
feedback here.
HLFIR dialect is a another example where conditional fast-math
support may be applied currently.

vzakhari committed:
[RFC][mlir] Conditional support for fast-math attributes.

This patch suggests changes for operations that support
arith::ArithFastMathInterface/LLVM::FastmathFlagsInterface.
Some of the operations may have fast-math flags not equal to `none`
only if they operate on floating point values.

This is inspired by https://llvm.org/docs/LangRef.html#fastmath-return-types
and my goal to add fast-math support for `arith.select` operation
that may produce results of any type.

The changes add new isArithFastMathApplicable/isFastmathApplicable
methods to the above interfaces that tell whether an operation
supporting the interface may have non-none fast-math flags.

LLVM dialect isFastmathApplicable implementation is based on https://github.com/llvm/llvm-project/blob/bac62ee5b473e70981a6bd9759ec316315fca07d/llvm/include/llvm/IR/Operator.h#L380
ARITH dialect isArithFastMathApplicable is more relaxed, because
it has to support custom MLIR types. This is the area where
improvements are needed (see TODO comments). I will appreciate
feedback here.
HLFIR dialect is a another example where conditional fast-math
support may be applied currently.

kuhar reviewed (commented):
> and my goal to add fast-math support for arith.select operation

Why would we want to have fast math flags over `arith.select`? What optimizations / rewrites does this allow?

vzakhari commented:
> > and my goal to add fast-math support for arith.select operation
> 
> Why would we want to have fast math flags over `arith.select`? What optimizations / rewrites does this allow?

For example, it enables vectorization of loops with min/max reductions in LLVM. Flang is currently producing `arith.select` without fast-math attrs.

In general, in LLVM any instruction that produces a floating point result may have fast-math flags. This includes FP PHIs and selects.

vzakhari commented on flang/lib/Optimizer/HLFIR/IR/HLFIRDialect.cpp:
```diff
@@ -237,3 +237,20 @@ bool hlfir::isFortranIntegerScalarOrArrayObject(mlir::Type type) {
+      }))
+    return true;
+
+  return true;
```
I will change this to `false` after fixing the lowering tests.

joker-eph commented on mlir/include/mlir/Dialect/Arith/IR/ArithOpsInterfaces.td:
```diff
@@ -22,31 +22,60 @@ def ArithFastMathInterface : OpInterface<"ArithFastMathInterface"> {
+    auto attr = fmi.getFastMathFlagsAttr();
+    if (attr && attr.getValue() != ::mlir::arith::FastMathFlags::none &&
+        !fmi.isArithFastMathApplicable())
+      return $_op->emitOpError() << "FastMathFlagsAttr is not applicable";
```
```suggestion
      return $_op->emitOpError() << "has flag " << stringify(attr.getValue()) << " but fast-math flags are not applicable (`isArithFastMathApplicable()` returns false)";
```

joker-eph commented on mlir/include/mlir/Dialect/Arith/IR/ArithOps.td:
```diff
@@ -1211,6 +1211,9 @@ def Arith_ExtFOp : Arith_FToFCastOp<"extf", [DeclareOpInterfaceMethods<ArithFast
     When operating on vectors, casts elementwise.
   }];
+  let extraClassDeclaration = [{
+    bool isApplicable() { return true; }
```
When is this method called?

Did you mean `isArithFastMathApplicable()` here? (if so we're missing a test to cover this)

joker-eph commented on flang/lib/Optimizer/CodeGen/CodeGen.cpp:
```diff
@@ -589,10 +589,18 @@ struct CallOpConversion : public fir::FIROpConversion<fir::CallOp> {
+    auto fmi =
+        mlir::cast<mlir::LLVM::FastmathFlagsInterface>(llvmCall.getOperation());
+    if (!fmi.isFastmathApplicable())
+      llvmCall->setAttr(
```
There should be better accessor on the LLVM::CallOp (this one is generic and quite expensive): ODS generates an accessor per-attribute.

chelini commented on mlir/include/mlir/Dialect/Arith/IR/ArithOpsInterfaces.td:
```diff
@@ -22,31 +22,60 @@ def ArithFastMathInterface : OpInterface<"ArithFastMathInterface"> {
+                         then the FastMathFlagsAttr of the operation
+                         must be nullptr or have 'none' value}],
+           /*returnType=*/"bool",
+           /*methodName=*/"isArithFastMathApplicable",
```
should I think of this as a sort of "verifier" for fastMath flags for the given operation?

chelini commented on mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:
```diff
@@ -225,6 +225,17 @@ def LLVM_FCmpOp : LLVM_ArithmeticCmpOp<"fcmp", [
+  let extraClassDeclaration = [{
+    /// Always allow FastmathFlags on llvm.fcmp.
+    /// It does not produce a floating point result, but
+    /// LLVM is currently relying on fast-math flags attached
```
do you mean that LLVM will look into the compare instead of the select operation for fastMath? Is unclear to me why you are changing this method for the cmpi and not the select.

vzakhari replied:
LLVM can look at both compare and select, depending on what it needs to do.

What I mean here is: LLVM's `fcmp` instruction supports fast-math flags, and `llvm.fcmp` operation should also support them; the general rule for instructions/operations to support fast-math is that they produce a floating point result; neither `fcmp` nor `llvm.fcmp` produce floating point result, so they are exceptions from the general rule, so `isFastmathApplicable` should be overridden here.

There is not need to override `isFastmathApplicable` for `llvm.select`, because it is covered by the general rule.

Note that the comment is explicitly saying that this is a temporary solution while LLVM expects it.

LLVM code has the following TODO about `fcmp`:
```
    // FIXME: To clean up and correct the semantics of fast-math-flags, FCmp
    //        should not be treated as a math op, but the other opcodes should.
    //        This would make things consistent with Select/PHI (FP value type
    //        determines whether they are math ops and, therefore, capable of
    //        having fast-math-flags).
```

kuhar commented:
> For example, it enables vectorization of loops with min/max reductions in LLVM. Flang is currently producing `arith.select` without fast-math attrs.
> 
> In general, in LLVM any instruction that produces a floating point result may have fast-math flags. This includes FP PHIs and selects.

Can you link to some example that shows why this is necessary? I'd think that `arith.select` preserves bitpatterns of its operands, so I struggle to see why it needs fast math flags; if llvm needs them, couldn't it calculate them as the intersection of the fast math flags of the operands?

vzakhari committed:
Addressed review comments.

dcaballe commented:
> > For example, it enables vectorization of loops with min/max reductions in LLVM. Flang is currently producing `arith.select` without fast-math attrs.
> > In general, in LLVM any instruction that produces a floating point result may have fast-math flags. This includes FP PHIs and selects.
> 
> Can you link to some example that shows why this is necessary? I'd think that `arith.select` preserves bitpatterns of its operands, so I struggle to see why it needs fast math flags; if llvm needs them, couldn't it calculate them as the intersection of the fast math flags of the operands?

I was asking myself the same question. I guess if an `arith.select` has `nnan` and one of the inputs is NaN we could turn it into poison...

kuhar commented:
> I was asking myself the same question. I guess if an `arith.select` has `nnan` and one of the inputs is NaN we could turn it into poison...

This could be supported by a dedicated unary op.

kuhar commented:
I think we could do something like this:
* add a new op, say `arith.assumef %x fastmath<nnan> : f32` whose sole purpose is to apply fast math flags its operand
* teach arith to llvm conversion to convert `arith.select(arith.assumef(x), arith.assumef(y))` to apply fast math flags to the produced `llvm.select` (and potentially other ops)
* teach the frontend to produce these `arith.assumef` ops when emitting selects with fast math enabled

vzakhari commented:
> I think we could do something like this:
> 
> * add a new op, say `%y = arith.assumef %x fastmath<nnan> : f32` (or `ub.assumef`?) whose sole purpose is to apply fast math flags to its operand
> * teach arith to llvm conversion to convert `arith.select(arith.assumef(x), arith.assumef(y))` to apply fast math flags to the produced `llvm.select` (and potentially other ops)
> * teach the frontend to produce these `arith.assumef` ops when emitting selects with fast math enabled

Can you please explain how this is better than having fast-math flags on the select itself?  It seems that what you are proposing depends on whether the operands' `arith.assume` definitions are always reachable from `arith.select`, but it may not be true with block arguments.  So having `arith.select` to carry fast-math seems more robust to me.

Please also note that this patch does not add fast-math support to `arith.select`, it just adds support for conditional fast-math support in general and for operations of HLFIR dialect.  I appreciate the discussion about `arith.select`, but I just want to understand if you are also proposing some changes to the current patch - can you please clarify?

vzakhari commented:
FYI, I addressed the review comments in https://github.com/llvm/llvm-project/commit/8834e36b29f50a3307de05c14bebe1f8ffa3bd1f, but github does not show it here for some reason.  I see `Processing updates` idle spin at the top of this page...

kuhar commented:
> Please also note that this patch does not add fast-math support to `arith.select`, it just adds support for conditional fast-math support in general and for operations of HLFIR dialect. I appreciate the discussion about `arith.select`, but I just want to understand if you are also proposing some changes to the current patch - can you please clarify?

Sorry, I was mostly thinking out loud and did not mean to derail this PR. I'm trying to understand the goal stated in the PR description:
> This is inspired by https://llvm.org/docs/LangRef.html#fastmath-return-types and my goal to add fast-math support for arith.select operation that may produce results of any type.

I'd not expect `arith.select` to know anything about fast math because of the reasons mentioned above in https://github.com/llvm/llvm-project/pull/125620#issuecomment-2634482765. Separately, fast math in clang/llvm is known to be broken and I'd like to make sure we are not working backwards from the solution and invertedly repeat the same mistakes. For example, see this issue: https://github.com/iree-org/iree/issues/19743. I will set aside some time to read the llvm discussion that you linked and understand their design better.

>  It seems that what you are proposing depends on whether the operands' arith.assume definitions are always reachable from arith.select, but it may not be true with block arguments. So having arith.select to carry fast-math seems more robust to me.

I'd think that the real implementation can have a matcher that checks if `arith.select` operands are `<insert-your-favorite-fmf-assumption>`, and frontends can insert these `assumef` ops to communicate this in the absence of anything else that would imply it.

vzakhari commented:
> Sorry, I was mostly thinking out loud and did not mean to derail this PR. I'm trying to understand the goal stated in the PR description:
> 
> > This is inspired by https://llvm.org/docs/LangRef.html#fastmath-return-types and my goal to add fast-math support for arith.select operation that may produce results of any type.

Thanks for the explanation!

> I'd think that the real implementation can have a matcher that checks if `arith.select` operands are `<insert-your-favorite-fmf-assumption>`, and frontends can insert these `assumef` ops to communicate this in the absence of anything else that would imply it.

I am not against the `assume` operations, but what I am trying to say is that MLIR region simplification may turn the operands of `arith.select` into block arguments, and so the matching will not be as easy as looking at the operands, it will be more as looking at all the possible values of the argument blocks and merging the fast-math flags meaning across all these values.  At the same time, some attribution of the block arguments with fast-math may be needed, anyway, if we decide to follow LLVM's approach and mark the PHIs with fast-math.

vzakhari commented on mlir/lib/Dialect/Arith/IR/ArithDialect.cpp:
```diff
@@ -66,3 +67,49 @@ Operation *arith::ArithDialect::materializeConstant(OpBuilder &builder,
+    return true;
+
+  // TODO: what about TupleType and custom dialect struct-like types?
+  // It seems that they worth an interface to get to the list of element types.
```
Any suggestions about this?

jeanPerier replied:
Do you have a use case of an operation with the fastmath interface that takes/returns such type?

If these operations are not common, maybe it is best/cheaper for the rest of the usages to keep the logic here simple and have these operation do the type visit as needed like you did in HLFIR.

That said, such type interface would make some sense to me in general.

vzakhari commented on mlir/lib/Dialect/Arith/IR/ArithDialect.cpp:
```diff
@@ -66,3 +67,49 @@ Operation *arith::ArithDialect::materializeConstant(OpBuilder &builder,
+/// TODO: the results often have the same type, and traversing
+/// the same type again and again is not very efficient.
+/// We can cache it here for the duration of the processing.
+/// Other ideas?
```
Any suggestions about this?

jeanPerier replied:
Are you talking about the result of a same operation with multiple results, or the result of different operation.

If this is about the former, it seems to ne the caching would be overkill given the average number or results in operation. If this is about the later, it seems to me that maintaining some shared cache somewhere would not be cheap, is the call to `isCompatibleType` that expensive?

jeanPerier commented on mlir/lib/Dialect/Arith/IR/ArithDialect.cpp:
```diff
@@ -66,3 +67,49 @@ Operation *arith::ArithDialect::materializeConstant(OpBuilder &builder,
+      return isCompatibleType(shapedType.getElementType());
+
+  // ComplexType's element type is always a FloatType.
+  if (auto complexType = dyn_cast<ComplexType>(type))
```
nit: move in isa with FloatType?

jeanPerier reviewed (approved):
Thanks @vzakhari, the flang dialect changes looks good and this makes sense to me in general. Please wait for the approval from those who had comments.

vzakhari commented:
> `arith.select` is a conditional move op and should never change the value of either operand.

I am not sure I understand how fast-math attributes attached to `arith.select` imply that it can change the value of either operand - can you please clarify?

> Ops like `arith.select` and block arguments are changed repeatedly throughout many pipelines and preserving (or even knowing how to set) fast math information is not feasible (or useful).

Isn't it true for any other discardable attribute?  In my opinion, the transformations should drop any attributes that they don't know about or preserve any attribute that they know about (generic attribute interfaces may ease the handling here).  Otherwise, how we can guarantee that a discardable attribute propagated during op->op transformation, without paying attention to preserving the attribute semantics, is still valid in the MLIR after the transformation?

I created https://discourse.llvm.org/t/rfc-arithfastmathinterface-support-for-arith-select/84508 for further discussion of the fast-math flags on `arith.select`. Can we please continue there?

I think the discussion here should be about conditional support of fast-math attributes on operations that support the fast-math interface.  I believe the HLFIR case is a good example where such support makes sence.

> If lowering into LLVM needs this information then it should perform analysis to determine the required information

I do not think it is always possible (see example in the above discourse RFC).

vzakhari reviewed (commented):
Thank you for the comments, Jean. I will upload updated files shortly.

vzakhari committed:
Addressed review comments.

vzakhari committed:
[flang] Fixed hlfir::isArithFastMathApplicable and LIT tests.
